File: lib/pacer/pipe/stream_uniq_pipe.rb
module Pacer::Pipes
  class StreamUniqPipe < RubyPipe
    def initialize(buffer = 1000)
      super()
      @list = java.util.LinkedList.new
      @buffer = buffer
    end

    protected

    def processNextStart
      while true
        obj = @starts.next
        duplicate = @list.removeLastOccurrence(obj)
        @list.addLast obj
        if not duplicate
          if @buffer == 0
            @list.removeFirst
          else
            @buffer -= 1
          end
          return obj
        end
      end
    rescue NativeException => e
      if e.cause.getClass == NoSuchElementException.getClass
        raise e.cause
      else
        raise e
      end
    end
  end
end

File: ext/src/main/java/com/xnlogic/pacer/pipes/StreamUniqPipe.java
package com.xnlogic.pacer.pipes;

import com.tinkerpop.pipes.AbstractPipe;
import java.util.LinkedList;

public class StreamUniqPipe<T> extends AbstractPipe<T, T> {
    private LinkedList list;
    private int buffer;
    
    public StreamUniqPipe(final int buffer) {
        super();
        this.list = new LinkedList();
        this.buffer = buffer;
    }

    public StreamUniqPipe() {
        this(1000);
    }

    protected T processNextStart() {
        while (true) {
            T obj = this.starts.next();
            boolean duplicate = this.list.removeLastOccurrence(obj);
            this.list.addLast(obj);

            if (!duplicate) {
                if (this.buffer == 0)
                    this.list.removeFirst();
                else
                    this.buffer--;

                return obj;
            }
        }
    }
}
